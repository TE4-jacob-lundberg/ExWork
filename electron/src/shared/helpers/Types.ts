export interface ILink {
  url: string;
  label: string;
}

export interface IGameFormFields {
  id: string;
  title: string;
  abbreviation: string;
  image: string; 
  xAxis: string;
  yAxis: string;
  links: ILink[];
  fileNames: string[];
}
