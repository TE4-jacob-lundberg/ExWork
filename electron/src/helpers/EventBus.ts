/* eslint-disable @typescript-eslint/no-explicit-any */
const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

export { EventBus };
