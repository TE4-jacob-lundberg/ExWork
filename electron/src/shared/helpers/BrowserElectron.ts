// eslint-disable-next-line @typescript-eslint/no-explicit-any
export let electron: any;

if (process.env.ENVIRONMENT === 'test') electron = require('electron');
else electron = window.require('electron');
