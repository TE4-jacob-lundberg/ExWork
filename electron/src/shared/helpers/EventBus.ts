/* eslint-disable @typescript-eslint/no-explicit-any */
interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

export { EventBus };
