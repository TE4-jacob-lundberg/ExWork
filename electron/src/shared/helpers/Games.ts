import { standardGames, IGame } from '../constants/standardGames';

export class Games {

  public static all(): IGame[] {
    if (!localStorage.getItem('games')) localStorage.setItem('games', JSON.stringify(standardGames));

    return JSON.parse(localStorage.getItem('games')!) || [];
  }

  public static addNew(game: IGame): string {
    const savedGames: IGame[] = JSON.parse(localStorage.getItem('games')!);
    if (savedGames.some(sGame => sGame.id === game.id)) return 'DUPLICATE_ID';
    
    localStorage.setItem('games', JSON.stringify(savedGames.concat(game)));
    return '';
  }
}
