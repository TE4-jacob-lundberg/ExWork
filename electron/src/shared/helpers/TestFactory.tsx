/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

export function factory(component: JSX.Element): RenderResult {
  return render(component);
}

export function routerFactory(component: JSX.Element, options: any = { history: createBrowserHistory() }): RenderResult {
  return factory(
    <Router history={options.history}>
      {component}
    </Router>
  );
}
