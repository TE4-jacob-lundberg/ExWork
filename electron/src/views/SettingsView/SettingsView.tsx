import React, { useEffect, useState, useCallback } from 'react';
import Styled from '@emotion/styled';
import { css } from '@emotion/core';
import { useIndexedDB } from 'react-indexed-db';

import { EventBus } from '../../shared/helpers/EventBus';
import ButtonComponent from '../../shared/components/ButtonComponent';
import PageTitleComponent from '../../shared/components/PageTitleComponent';
import InputSettingComponent from './components/InputSettingComponent';
import { IKeybind } from '../../shared/helpers/Types';

const ContainerStyled = Styled.div`
  height: 100%;
  width: 100%;
`;

const SettingContainer = Styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin: 16px 0;
  width: 100%;
`;

const LabelStyled = Styled.h1`
  font-size: 1rem;
  margin: 8px;
`;

const SettingSectionStyling = css`
  width: 75%;
  margin: 32px auto;
  padding: 20px;
  display: flex;
  flex-direction: column;
  align-items: center;
  border-radius: 5px;
`;

const GeneralContainerStyled = Styled.div`
  ${SettingSectionStyling.styles}
  border: 4px solid #555;
  background: #333;
`;

const DangerContainerStyled = Styled.div`
  ${SettingSectionStyling.styles}
  border: 4px solid red;
  background: #d43939;

  & label {
    text-transform: uppercase;
  }
`;  

interface Props {}

const SettingsView: React.FC<Props> = function () {
  const [keybinds, setKeybinds] = useState<IKeybind[]>();
  const gameDB = useIndexedDB('games');
  const keybindDB = useIndexedDB('keybinds');
  
  const getInputs = useCallback((): void => {
    keybindDB.getAll().then((resp: IKeybind[]) => setKeybinds(resp));
  }, [keybindDB]);
  
  useEffect(() => {
    if (keybinds) return;
    getInputs();
  }, [keybinds, getInputs]);

  useEffect(() => EventBus.subscribe('new-keybinds-set', getInputs), [getInputs]);

  return (
    <>
      <PageTitleComponent title="Settings" />
      <ContainerStyled>
        <GeneralContainerStyled>
          {keybinds && keybinds.map(keybind => (
            <SettingContainer key={keybind.name}>
              <LabelStyled>{keybind.name}</LabelStyled>
              <InputSettingComponent config={keybind} />
            </SettingContainer>
          ))}
        </GeneralContainerStyled>
        <DangerContainerStyled>
          <SettingContainer>
            <LabelStyled>Reset keybinds to default</LabelStyled>
            <ButtonComponent
              background="red"
              onClick={(): Promise<void> => keybindDB.clear().then(() => EventBus.dispatch('new-keybinds'))}
              confirm
            >
              RESET KEYBINDS
            </ButtonComponent>
          </SettingContainer>
          <SettingContainer>
            <LabelStyled>Reset your games to when first opened</LabelStyled>
            <ButtonComponent
              background="red"
              onClick={(): Promise<void> => gameDB.clear().then(() => EventBus.dispatch('add-standard-games'))}
              confirm
            >
              RESET GAMES
            </ButtonComponent>
          </SettingContainer>
        </DangerContainerStyled>
      </ContainerStyled>
    </>
  );
};

export default SettingsView;
