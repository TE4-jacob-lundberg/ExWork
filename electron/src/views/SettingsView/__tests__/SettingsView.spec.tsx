import React from 'react';
import { fireEvent, waitForElement } from '@testing-library/react';

import { factory } from '../../../shared/helpers/TestFactory';
import SettingsView from '../SettingsView';
import { EventBus } from '../../../shared/helpers/EventBus';
import { IKeybind } from '../../../shared/helpers/Types';

jest.mock('../../../shared/helpers/EventBus');

jest.mock('react-indexed-db', () => ({
  useIndexedDB: (): object => {
    return {
      clear: (): Promise<void> => Promise.resolve(),
      getAll: (): Promise<IKeybind[]> => Promise.resolve([
        { name: 'Test Keybind', key: 'F10'},
      ]),
    };
  },
}));

const testComponent = <SettingsView />;

describe('SettingsView.tsx', () => {
  it('renders correctly', () => {
    const { container } = factory(testComponent);
    
    expect(container).toMatchSnapshot();
  });

  it('allows resetting games', async () => {
    const { getByText, queryByText } = factory(testComponent);

    fireEvent.click(getByText('RESET GAMES'));
    await waitForElement(() => getByText('Are you sure?'));

    fireEvent.click(getByText('Continue'));
    await waitForElement(() => !queryByText('Are you sure?'));

    expect(EventBus.dispatch).toHaveBeenCalledTimes(1);
  });
});
