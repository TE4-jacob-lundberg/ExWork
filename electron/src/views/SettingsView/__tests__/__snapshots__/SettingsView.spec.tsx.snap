// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsView.tsx renders correctly 1`] = `
<div>
  <div
    class="css-j99gk8"
    data-testid="page-title"
  >
    <h1
      class="css-p1pfmz"
    >
      Settings
    </h1>
  </div>
  <div
    class="css-olgc7d"
  >
    <div
      class="css-1iagirt"
    />
    <div
      class="css-4ng0sz"
    >
      <div
        class="css-jt7ygj"
      >
        <h1
          class="css-1sx427h"
        >
          Reset keybinds to default
        </h1>
        <button
          class="css-34a7o7"
        >
          RESET KEYBINDS
        </button>
      </div>
      <div
        class="css-jt7ygj"
      >
        <h1
          class="css-1sx427h"
        >
          Reset your games to when first opened
        </h1>
        <button
          class="css-34a7o7"
        >
          RESET GAMES
        </button>
      </div>
    </div>
  </div>
</div>
`;
